cranelift/codegen/src/machinst/isle.rs

```diff
@@ -585,9 +585,50 @@ macro_rules! isle_lower_prelude_methods {
                 .collect();
             self.lower_ctx.gen_return(rets);
         }
+
+        fn shuffle32_from_imm(&mut self, imm: Immediate) -> Option<(u8, u8, u8, u8)> {
+            let bytes = self.lower_ctx.get_immediate_data(imm).as_slice();
+            Some((
+                crate::machinst::isle::shuffle_imm(4, &bytes[0..4])?,
+                crate::machinst::isle::shuffle_imm(4, &bytes[4..8])?,
+                crate::machinst::isle::shuffle_imm(4, &bytes[8..12])?,
+                crate::machinst::isle::shuffle_imm(4, &bytes[12..16])?,
+            ))
+        }
+
+        fn shuffle16_from_imm(
+            &mut self,
+            imm: Immediate,
+        ) -> Option<(u8, u8, u8, u8, u8, u8, u8, u8)> {
+            let bytes = self.lower_ctx.get_immediate_data(imm).as_slice();
+            Some((
+                crate::machinst::isle::shuffle_imm(2, &bytes[0..2])?,
+                crate::machinst::isle::shuffle_imm(2, &bytes[2..4])?,
+                crate::machinst::isle::shuffle_imm(2, &bytes[4..6])?,
+                crate::machinst::isle::shuffle_imm(2, &bytes[6..8])?,
+                crate::machinst::isle::shuffle_imm(2, &bytes[8..10])?,
+                crate::machinst::isle::shuffle_imm(2, &bytes[10..12])?,
+                crate::machinst::isle::shuffle_imm(2, &bytes[12..14])?,
+                crate::machinst::isle::shuffle_imm(2, &bytes[14..16])?,
+            ))
+        }
     };
 }
 
+pub fn shuffle_imm(size: u8, bytes: &[u8]) -> Option<u8> {
```

Comment:
I think this could use some documentation; it's not exactly clear why only `bytes[0]` matters here (little-endianness-related, it would seem). This is a foundational block to all of the helpers above and it would be nice to see clearly when we expect the `None`s from this function to be the ones propagated outwards.